flopy/mf6/data/mfdataplist.py

```diff
@@ -0,0 +1,2527 @@
+            if not (isinstance(data[0], list) or isinstance(data[0], tuple)):
+                # get data in the format of a tuple of lists (or tuples)
+                data = [data]
+            # resolve the data's column headings
+            columns = self._resolve_columns(data)[0]
+            if columns is None:
+                message = (
+                    f"ERROR: Data list {self._data_name} supplied the "
+                    f"wrong number of columns of data, expected "
+                    f"{len(self._data_item_names)} got {len(data[0])}."
+                )
+                type_, value_, traceback_ = sys.exc_info()
+                raise MFDataException(
+                    self._data_dimensions.structure.get_model(),
+                    self._data_dimensions.structure.get_package(),
+                    self._data_dimensions.structure.path,
+                    "setting list data",
+                    self._data_dimensions.structure.name,
+                    inspect.stack()[0][3],
+                    type_,
+                    value_,
+                    traceback_,
+                    message,
+                    self._simulation_data.debug,
+                )
+
+            if len(data[0]) == 0:
+                # create empty dataset
+                data = pandas.DataFrame(columns=columns)
+            else:
+                # create dataset
+                data = pandas.DataFrame(data, columns=columns)
+            if (
+                self._data_item_names[-1] == "boundname"
+                and "boundname" not in columns
+            ):
+                # add empty boundname column
+                data["boundname"] = ""
+            # get rid of tuples from cellids
+            data, count = self._untuple_cellids(data)
+            if count > 0:
+                # make sure columns are still in correct order
+                data = pandas.DataFrame(data, columns=self._header_names)
+        elif isinstance(data, pandas.DataFrame):
+            if len(data.columns) != len(self._header_names):
+                message = (
+                    f"ERROR: Data list {self._data_name} supplied the "
+                    f"wrong number of columns of data, expected "
+                    f"{len(self._data_item_names)} got {len(data[0])}."
+                )
+                type_, value_, traceback_ = sys.exc_info()
+                raise MFDataException(
+                    self._data_dimensions.structure.get_model(),
+                    self._data_dimensions.structure.get_package(),
+                    self._data_dimensions.structure.path,
+                    "setting list data",
+                    self._data_dimensions.structure.name,
+                    inspect.stack()[0][3],
+                    type_,
+                    value_,
+                    traceback_,
+                    message,
+                    self._simulation_data.debug,
+                )
```

Comment:
I can not easily determine the name of the missing column since this code accepts Pandas Dataframes with incorrect column names (it corrects the column names below, which is trivial to do given that the Dataframe has the correct number of columns).  I did however make a more detailed error message that lists the data column names supplied and expected.